SmartDeviceLink/SDLEncryptionLifecycleManager.m

```diff
@@ -0,0 +1,247 @@
+    SDLLogD(@"Encryption manager is ready");
+}
+
+- (void)didEnterStateEncryptionStopped {
+    SDLLogD(@"Encryption manager stopped");
+}
+
+#pragma mark - SDLProtocolListener
+#pragma mark Encryption Start Service ACK
+
+- (void)handleProtocolStartServiceACKMessage:(SDLProtocolMessage *)startServiceACK {
+    switch (startServiceACK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            [self sdl_handleEncryptionStartServiceACK:startServiceACK];
+        } break;
+        default: break;
+    }
+}
+
+- (void)sdl_handleEncryptionStartServiceACK:(SDLProtocolMessage *)encryptionStartServiceAck {
+    if (encryptionStartServiceAck.header.encrypted) {
+        SDLLogD(@"Encryption service started");
+        [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateReady];
+    } else {
+        SDLLogD(@"Encryption service ACK received encryption = OFF");
+        [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+    }
+}
+
+#pragma mark Encryption Start Service NAK
+
+- (void)handleProtocolStartServiceNAKMessage:(SDLProtocolMessage *)startServiceNAK {
+    switch (startServiceNAK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            if (startServiceNAK.header.encrypted) {
+                [self sdl_handleEncryptionStartServiceNAK:startServiceNAK];
+            } else {
+                SDLLogW(@"Encryption service failed to start due to encryption bit set to 0 in ACK");
+            }
+        } break;
+        default: break;
+    }
+}
+
+- (void)sdl_handleEncryptionStartServiceNAK:(SDLProtocolMessage *)audioStartServiceNak {
+    SDLLogW(@"Encryption service failed to start due to NAK");
+    [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+}
+
+#pragma mark Encryption End Service
+
+- (void)handleProtocolEndServiceACKMessage:(SDLProtocolMessage *)endServiceACK {
+    switch (endServiceACK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            SDLLogW(@"Encryption RPC service ended with end service ACK");
+            [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+        } break;
+        default: break;
+    }
+}
+
+- (void)handleProtocolEndServiceNAKMessage:(SDLProtocolMessage *)endServiceNAK {
+    switch (endServiceNAK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            SDLLogW(@"Encryption RPC service ended with end service NAK");
+            [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+        } break;
+        default: break;
+    }
+}
+
+#pragma mark - SDL RPC Notification callbacks
+- (void)sdl_permissionsDidChange:(SDLRPCNotificationNotification *)notification {
+    if (![notification isNotificationMemberOfClass:[SDLOnPermissionsChange class]]) {
+        return;
+    }
+    
+    SDLOnPermissionsChange *onPermissionChange = notification.notification;
```

Comment:
Check `onPermissionChange.requireEncryption` here to see if it `false` and return early after emptying the `permissions` dictionary. The individual permissions only need to be checked if `requireEncryption` is `nil` or `true`.

Reply:
Sounds good, made th changes